runtime/configurationparser/ConfigurationParser.cpp

```diff
@@ -16,65 +17,89 @@ extern "C" {
   }
 }
 
-static void *allocatePatternAsConfiguration(const KOREPattern *Pattern) {
-  const auto constructor = dynamic_cast<const KORECompositePattern *>(Pattern);
-  assert(constructor);
+struct construction {
+  uint32_t tag;
+  size_t nchildren;
+};
 
-  const KORESymbol *symbol = constructor->getConstructor();
-  assert(symbol->isConcrete() && "found sort variable in initial configuration");
-  if (symbol->getName() == "\\dv") {
-    const auto sort = dynamic_cast<KORECompositeSort *>(symbol->getFormalArguments()[0].get());
-    const auto strPattern =
-      dynamic_cast<KOREStringPattern *>(constructor->getArguments()[0].get());
-    std::string contents = strPattern->getContents();
-    return getToken(sort->getName().c_str(), contents.size(), contents.c_str());
-  }
-  std::ostringstream Out;
-  symbol->print(Out);
-  uint32_t tag = getTagForSymbolName(Out.str().c_str());
+static void *constructInitialConfiguration(const KOREPattern *initial) {
+  std::vector<std::variant<const KOREPattern *, construction>> workList{initial};
+  std::vector<void *> output;
+  while (!workList.empty()) {
+    std::variant<const KOREPattern *, construction> current = workList.back();
+    workList.pop_back();
+    if (current.index() == 0) {
+      const auto constructor = dynamic_cast<const KORECompositePattern *>(std::get<const KOREPattern *>(current));
+      assert(constructor);
 
-  if (isSymbolAFunction(tag)) {
-    std::vector<void *> arguments;
-    for (const auto &child : constructor->getArguments()) {
-      arguments.push_back(allocatePatternAsConfiguration(child.get()));
-    }
-    return evaluateFunctionSymbol(tag, &arguments[0]);
-  }
+      const KORESymbol *symbol = constructor->getConstructor();
+      assert(symbol->isConcrete() && "found sort variable in initial configuration");
+      if (symbol->getName() == "\\dv") {
+        const auto sort = dynamic_cast<KORECompositeSort *>(symbol->getFormalArguments()[0].get());
+        const auto strPattern =
+          dynamic_cast<KOREStringPattern *>(constructor->getArguments()[0].get());
+        std::string contents = strPattern->getContents();
+        output.push_back(getToken(sort->getName().c_str(), contents.size(), contents.c_str()));
+        continue;
+      }
 
-  struct blockheader headerVal = getBlockHeaderForSymbol(tag);
-  size_t size = size_hdr(headerVal.hdr);
-  
-  if (size == 8) {
-    return (block *) ((uint64_t)tag << 32 | 1);
-  }
+      std::ostringstream Out;
+      symbol->print(Out);
+      uint32_t tag = getTagForSymbolName(Out.str().c_str());
+      if (isSymbolAFunction(tag) && constructor->getArguments().empty()) {
+        output.push_back(evaluateFunctionSymbol(tag, nullptr));
+        continue;
+      } else if (constructor->getArguments().empty()) {
+        output.push_back((block *) ((uint64_t)tag << 32 | 1));
```

Comment:
I'd add a function/macro whose name explains what kind of encoding is done here.